### ISSUE TITLE
CDN caching issues with 404s from asset bundle during upgrade

### ISSUE THREAD

**Issue by sentientmonkey** [argoproj/argo-cd#15084]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

We recently upgraded from argocd v2.6.7 to v2.8.0. During our upgrade, our argocd ui loaded with a blank page - checking the javascript error console, we saw we were getting "X-Content-Type-Options" errors from trying to load the manifest bundle as "text/plain" instead of javascript. On further analysis, we saw that while our pods were rolling, new pods were looking for the new asset bundle file and were returning 404s from the old pods. This would have been ok, but we found that since we are using CloudFlare's proxy, the 404s were cached for up to 3m after our deployment finished.

https://developers.cloudflare.com/cache/how-to/configure-cache-status-code

We were able to purge the cache for the new asset bundle, but it some digging to reproduce. We can likely set up a cache rule in cloudflare, but I would expect this to cause issues with anyone that runs argocd with a caching layer.

**To Reproduce**

1. Deploy an older argocd version (i.e. 2.6.7)
2. Use Cloudflare or other cache to proxy requests
3. Upgrade to newer version of argocd (i.e. 2.8.0)

**Expected behavior**

Ideally, for the asset bundles when they return a 404, they'd return a `Cache-Control: no-cache` header in the response to avoid caching the 404s. This would mitigate any caching that happen during a deployment. Since the cache headers are already set for the bundle properly as `public, max-age=31536000, immutable` this would allow rolling deploys without any disruption if you have a proxy or CDN that front the argocd instance.

**Screenshots**
n/a

**Version**

```shell
argocd: v2.8.0+804d4b8
  BuildDate: 2023-08-07T14:25:33Z
  GitCommit: 804d4b8ca6bc4c2cf02c5c971aa923ec5b8623f0
  GitTreeState: clean
  GoVersion: go1.20.6
  Compiler: gc
  Platform: linux/amd64
```

**Logs**

```
n/a
```

**Comment by sentientmonkey** [comment]
This also appears similar to the issue brought up for badge statuses https://github.com/argoproj/argo-cd/pull/2328

**Comment by crenshaw-dev** [comment]
> Ideally, for the asset bundles when they return a 404, they'd return a Cache-Control: no-cache header in the response to avoid caching the 404s.

Sounds good to me. Are you up for a PR?

**Comment by sentientmonkey** [comment]
@crenshaw-dev I've added a PR for this. Let me know if you have any feedback/questions! I backfilled some unit tests along the way.

**Comment by blakepettersson** [comment]
Fixed with #15327.
